dbms/src/Flash/Pipeline/Schedule/Event/PlainPipelineEvent.cpp

```diff
@@ -0,0 +1,37 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Pipeline/Pipeline.h>
+#include <Flash/Pipeline/Schedule/Event/PlainPipelineEvent.h>
+#include <Flash/Pipeline/Schedule/Task/PipelineTask.h>
+
+namespace DB
+{
+bool PlainPipelineEvent::scheduleImpl()
+{
+    assert(pipeline);
+    auto pipline_exec_group = pipeline->toExecGroup(context, concurrency);
+    pipeline.reset();
```

Comment:
Why release the pipeline here?

Reply:
Because we should release the pipeline before `exec_status.completeEvent()`, otherwise it may cause Pipeline to be destructed later than MPPTask.
And after `pipeline->toExecGroup(context, concurrency);`, `pipeline` is useless here.

Reply:
moved to `PipelineEvent::finishImpl`

Reply:
It will be nice to add some comments about what must be done and only done in `finishImpl`, i.e., releasing resources that must be released ahead of event destruction.

Reply:
ok, has added in `Event::finishImpl` and `PipelineEvent::finishImpl`